python/ray/data/_internal/block_batching/block_batching.py

```diff
@@ -144,84 +143,43 @@ def batch_blocks(
     This means that this function does not support block prefetching.
     """
 
-    batch_iter = _format_batches(
-        _blocks_to_batches(
-            block_iter=blocks,
+    def _iterator_fn(base_iterator: Iterator[Block]) -> Iterator[DataBatch]:
+        batch_iter = _format_batches(
+            _blocks_to_batches(
+                block_iter=base_iterator,
+                stats=stats,
+                batch_size=batch_size,
+                drop_last=drop_last,
+                shuffle_buffer_min_size=shuffle_buffer_min_size,
+                shuffle_seed=shuffle_seed,
+                ensure_copy=ensure_copy,
+            ),
+            batch_format=batch_format,
             stats=stats,
-            batch_size=batch_size,
-            drop_last=drop_last,
-            shuffle_buffer_min_size=shuffle_buffer_min_size,
-            shuffle_seed=shuffle_seed,
-            ensure_copy=ensure_copy,
-        ),
-        batch_format=batch_format,
-        stats=stats,
-    )
+        )
 
-    if collate_fn is not None:
+        if collate_fn is not None:
 
-        def batch_fn_iter(iterator: Iterator[DataBatch]) -> Iterator[DataBatch]:
-            for batch in iterator:
-                yield collate_fn(batch)
+            def batch_fn_iter(iterator: Iterator[DataBatch]) -> Iterator[DataBatch]:
+                for batch in iterator:
+                    yield collate_fn(batch)
 
-        batch_iter = batch_fn_iter(batch_iter)
+            batch_iter = batch_fn_iter(batch_iter)
+        yield from batch_iter
```

Comment:
Hmm so this wrapping `_iterator_fn` is necessary because in `_make_async_gen` we want to synchronize the `next(blocks)` but _not_ the `next(batch_iter)`, since the latter is the expensive bit (batching, formatting) that we want to parallelize? Might want to leave that as a comment to make that clear.

Reply:
Yep that's right!

Reply:
I actually think the current approach of passing in a separate `fn` to execute is more intuitive.

Reply:
Agreed, it's definitely less ambiguous than trying to determine in which thread the implicit `next()` is being called, e.g. when iterating over an iterator via a for loop.

Reply:
added comment